src/guiapi/src/window.cpp

```diff
@@ -4,9 +4,95 @@
 #include "gui.hpp"
 #include <algorithm> // std::find
 #include "ScreenHandler.hpp"
+#include "dbg.h"
 
 extern osThreadId displayTaskHandle;
 
+constexpr bool GUI_event_IsKnob(GUI_event_t event) {
+    switch (event) {
+    case GUI_event_t::BTN_DN:
+    case GUI_event_t::BTN_UP:
+    case GUI_event_t::ENC_DN:
+    case GUI_event_t::ENC_UP:
+        return true;
+    default:
+        return false;
+    }
+}
+
+constexpr bool GUI_event_IsWindowKnobReaction(GUI_event_t event) {
+    switch (event) {
+    case GUI_event_t::FOCUS0:
+    case GUI_event_t::FOCUS1:
+    case GUI_event_t::CAPT_0:
+    case GUI_event_t::CAPT_1:
+    case GUI_event_t::CLICK:
+    case GUI_event_t::DOUBLE_CLICK:
+    case GUI_event_t::HOLD:
+    case GUI_event_t::CHANGE:
+        return true;
+    default:
+        return false;
+    }
+}
+
+constexpr bool GUI_event_IsAnyButLoop(GUI_event_t event) {
+    return event != GUI_event_t::LOOP;
+}
+
+constexpr const char *GUI_event_prt(GUI_event_t event) {
+    switch (event) {
+    case GUI_event_t::BTN_DN:
+        return ("button down");
```

Comment:
Are you really returning an address of a local variable?
Technically, this may work - the string must be somewhere in CPU FLASH and the compiler might be smart enough to not-to load it into RAM and just return the address in FLASH, but the whole function is highly suspicios.

Reply:
cannot do it in constexpr function